url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/dnssec/DNSSECWorld.java

```diff
@@ -0,0 +1,370 @@
+            dos.write(toUnsignedByteArray(privateKey.getParams().getP(), t * 8 + 64));
+            dos.write(toUnsignedByteArray(privateKey.getParams().getG(), t * 8 + 64));
+            dos.write(toUnsignedByteArray(y, t * 8 + 64));
+            return baos.toByteArray();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static byte[] getRSAPublicKey(RSAPrivateCrtKey privateKey) {
+        try {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            DataOutputStream dos = new DataOutputStream(baos);
+            byte[] exponent = toUnsignedByteArray(privateKey.getPublicExponent());
+            if (exponent.length > 255) {
+                dos.writeByte(0);
+                dos.writeShort(exponent.length);
+            } else {
+                dos.writeByte(exponent.length);
+            }
+            dos.write(exponent);
+            dos.write(toUnsignedByteArray(privateKey.getModulus()));
+            return baos.toByteArray();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static byte[] toUnsignedByteArray(BigInteger bigInteger) {
+        byte[] array = bigInteger.toByteArray();
+        if (array[0] == 0) {
+            byte[] tmp = new byte[array.length - 1];
+            System.arraycopy(array, 1, tmp, 0, tmp.length);
+            array = tmp;
+        }
+        return array;
+    }
+
+    private static byte[] toUnsignedByteArray(BigInteger bigInteger, int length) {
+        byte[] array = bigInteger.toByteArray();
+        if (array.length != length) {
+            if (array.length == length + 1 && array[0] == 0) {
+                byte[] tmp = new byte[array.length - 1];
+                System.arraycopy(array, 1, tmp, 0, tmp.length);
+                array = tmp;
+            } else if (array.length < length) {
+                byte[] tmp = new byte[length];
+                System.arraycopy(array, 0, tmp, length - array.length, array.length);
+                array = tmp;
+            }
+        }
+        return array;
+    }
+
+    public static class AddressedNsecResponse implements PreparedResponse {
+        final InetAddress address;
+        final DNSMessage nsecMessage;
+
+        public AddressedNsecResponse(InetAddress address, DNSMessage nsecMessage) {
+            this.address = address;
+            this.nsecMessage = nsecMessage;
+        }
+
+        @Override
+        public boolean isResponse(DNSMessage request, InetAddress address) {
+            Record<? extends Data> nsecRecord = null;
+            for (Record<? extends Data> record : nsecMessage.authoritySection) {
+                if (record.type == Record.TYPE.NSEC)
+                    nsecRecord = record;
+            }
+            return address.equals(this.address) && Verifier.nsecMatches(request.getQuestion().name.ace, nsecRecord.name.ace, ((NSEC) nsecRecord.payloadData).next.ace);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;request&#39; (index: 0 | type: DNSMessage). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;address&#39;. The same parameter type is validated here for example: url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/AbstractDNSClient.java:147. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).